ipaserver/plugins/cert.py

```diff
@@ -1643,21 +1643,26 @@ def _ca_search(self, raw, pkey_only, exactly, **options):
         result = collections.OrderedDict()
         complete = bool(ra_options)
 
-        # workaround for RHBZ#1669012
-        # Improve performance for service and host case by also searching
-        # for subject. This limits the amount of certificate retrieved from
-        # Dogtag. The special case is only used, when no ra_options are set
-        # and exactly one service or host is supplied.
-        # The complete flag is left to False.
+        # workaround for RHBZ#1669012 and RHBZ#1695685
+        # Improve performance for service, host and user case by also
+        # searching for subject. This limits the amount of certificate
+        # retrieved from Dogtag. The special case is only used, when
+        # no ra_options are set and exactly one service, host, or user is
+        # supplied.
+        # IPA enforces that subject CN is either a hostname or a username.
+        # The complete flag is left to False to catch overrides.
         if not ra_options:
             services = options.get('service', ())
             hosts = options.get('host', ())
-            if len(services) == 1 and not hosts:
-                principal = kerberos.Principal(options['service'][0])
+            users = options.get('users', ())
```

Comment:
This should be `options.get('user', ())` (the key in `options` is `'user'`).

However with this correction in place, it is clear that the code regresses in the subject DN corner case I [mentioned above](https://github.com/freeipa/freeipa/pull/2990/files#r272901851).

So... I am considering NAKing this.  We might need to go back to the drawing board on `cert-find`.

OTOH the scenario that breaks this code is a corner case.  It's not the end of the world if we ship this patch as-is.  `cert-find` has a lot of limitations as it is, and this patch may help in the common case.

Reply:
Thanks for the view and good catch! I have updated the patch.

Do we already have a ticket to track the comments? GH has a tendency to eat comments on rebased PRs.

Reply:
@tiran I filed https://pagure.io/freeipa/issue/7904 to capture the comments and discuss re-working cert-find and friends.